Add auto-refresh function for Nexus files.

TerryGeng opened the pull request:
In some cases, we'd like to utilize the SWMR feature of HDF5 files and reload the file on the fly. This commit implements this feature.

jkotan commented:
Hello @TerryGeng ,

we use the SWMR  mode with dedicated for this purpose the "NeXus File" image source. 
In the Lavue Configuration you have to options for it:

    Keep the file open:
    Start from the last image:

Have you try to use it?

TerryGeng commented:
I'm sorry that I didn't see that option. It hides behind the expert mode :)

But still, with these two options on, still I have to manually click the "next" button or drag the slider to see more frames have been added to the file.

One feature of this PR is once I check the "auto-refresh" box, lavue will periodically reload the file so I can see the slider is growing. And if I type "-1" into the frame selection box, the frame plotted will always be updated with the last frame of the data entry.

jkotan commented:
No, it is in a different place, i.e. the left upper corner (you don't use the Load button):

https://confluence.desy.de/display/FSEC/LaVue+-+NeXus+File

Like for other image sources you press the Start button and the image view is updated when new data appear in the file.
So it works exactly the same as your proposed  auto-refresh widget.

TerryGeng commented:
Yes, I did notice this part. But here's the tricky thing: if I open the image file with the Load button, I can't get the file automatically refreshed, if I use the Source panel, I can get the file updated, but I can't see the previous frames. :P

jkotan reviewed (changes requested):
You are right. I have to test your PR more. 

- It looks like you remove loading metadata from nexus files. Is it by purpose? 
- PR mess up loading file. With the PR when another file was loaded before you cannot load a new file anymore.

TerryGeng commented:
Thanks for the quick response! I have fixed the file loading bug just now.

As for the metadata, I notice this variable is just used for printing metadata to the console. I thought this is just for debugging purposes. I remove it since this metadata will be printed every time the file is refreshed and has flooded my console.

Now I have uncommented those lines.

jkotan commented:
Dear @TerryGeng, I discuss your PR with colleges  at DESY. It seems than your PR introduces unwanted interference between image sources mechanism and  your image auto-re-loader. When they run together we don't know from where a new image comes from. 
Therefore to satisfy you requirements it is better to add possibility of accessing past images in NeXus File image source. Using "Memory Buffer" widget it is possible to view past images from any image source. But if you wish we can add widget to select  a particulate image frame in the NeXus File image source.
For loading imaging I think it is enough to add reload button to grep the new image frame in open nexus file.
What do you think about the proposition?

TerryGeng commented:
Hi @jkotan, it is great to know you are actively working on this idea and have come up with these proposals. Actually, for us, the motivation of tweaking LaVue is to make LaVue a part of our experiment framework at BSRF. (Thanks for making such a good tool!)
I have noticed that the procedure of loading images in LaVue is complicated and subtle, and I'm sorry that I wasn't able to thoroughly examine all the code to avoid making buggy commits.

I think your proposals make sense to me. For our use case, we won't choose to save past images in the memory since the images are huge. In that case, I think it'd be better to have a spinbox or something on the NeXus File image source widget, then pass it as part of the configuration to the image source.

I think I'm going to work on this idea later. Thanks again for your useful advice.

TerryGeng committed:
Add a spinbox for frame selection in NXSFileSourceWidget, removed "last frame" option in the expert configuration.

TerryGeng commented:
Hi, I have implemented your proposal. I noticed that after I added this "frame" spinbox in the source widget, the "open from the last frame" option in the expert configuration is duplicated so I removed it.

TerryGeng commented on lavuelib/imageSource.py:
```diff
@@ -361,43 +359,31 @@ def getData(self):
                     #     print("IMAGE Metadata = %s" % str(metadata))
                 fid = self.__handler.getLastFrame(self.__node, self.__gdim)
-                if self.__nxslast:
-                    if fid - 1 > self.__frame or fid - 1 < self.__frame:
```
Why -1? I think getLastFrame returns just the index of the last frame, not the count of the dataset.

jkotan replied:
Hello @TerryGeng, 
Thank you for your PR. I have a few comment on it.
- "open from the last frame" have to state as it is. It was requested by our users. With this feature you can "play the  movie" of all your images from the first one to the last one. I would expect that the fame variable you add would represent the last frame to show  in the movie when the feature is on. If the feature is off  and start button in on it shows
the frame selected in the new widget. As you have noticed minus indices denote frames from the end so frame=-1 would correspond the the behavior without your PR.
- it would be good to have possibility to change frames after the "Start" button is pressed.
- with `getLastFrame` in principle you are right but the current indices inform us also when there is no frame in the image (returns 0). Therefore, it is better to change the method name to something like getNumberOfFrames or getFrameCounts.

TerryGeng commented on lavuelib/imageSource.py:
```diff
@@ -361,43 +359,31 @@ def getData(self):
                     self.__node, self.__frame, self.__gdim)
-            # except Exception:
             except Exception:
-                try:
```
This part of the code is largely duplicated with L364 to L383. I found the only difference between these two parts is about the metadata, so I refactored it.

jkotan replied:
Yes, you are right. Thanks.

TerryGeng commented on lavuelib/sourceWidget.py:
```diff
@@ -1752,6 +1750,7 @@ def configure(self, configuration):
         :param configuration: configuration string
         :type configuration: :obj:`str`
         """
+        print(configuration)
```
Interestingly this part of code is never called in my test. And the format of the configuration indicated in this part of code is different from what it is in the other parts. I assume this part is obsolete and perhaps needs to be removed in the future.

jkotan replied:
it is called from "LiveViewer.__applyoptions()" in two cases:
-  when you start lavue with -c option e.g.
       lavue -s nxsfile -c /tmp/myfile.nxs://entry/data/pilatus,1
- when you use Tango Control System (Epics is not supported in this case)  you can set a LavueController  server and change lavue configuration with its LavueController.LavueState attrubute.

A number in the configuration above corresponds to a staking dimension of the dataset. We can add a second numberfor the frame number you added.

TerryGeng commented on lavuelib/ui/ConfigDialog.ui:
```diff
@@ -1485,23 +1475,10 @@
                 </property>
                </widget>
               </item>
-              <item row="1" column="0">
```
Remove the old checkbox.

jkotan committed:
Merge branch 'udevelop' into develop

jkotan committed:
Merge remote-tracking branch 'upstream/develop' into develop

jkotan committed:
add nxslast option back

jkotan committed:
change allowed frame range in nexus file image source

jkotan committed:
apply pep8

jkotan committed:
fix passing parameters to Nexus File image source

jkotan committed:
fix passing frame numver for nxslast=False

jkotan commented:
Hello @TerryGeng,

Since we would like to merge your PR before the end of our summer shutdown and you looks like on summer holidays I apply the above comments to your PR

TerryGeng commented:
Hi! Thanks for helping me with the rest of the work and getting this PR merged!

Actually I was stuck at trying to implement the "changing frames on the fly" thing. It looks like this can't be achieved without some major adjustment to the structure of the program.

I think I will look into this in the future, when I get some time.

Anyway, thanks for your help!
